cmd/vocab-service/main.go

```diff
@@ -0,0 +1,200 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/rs/cors"
+	"github.com/tylerb/graceful"
+)
+
+type rpcRequest struct {
+	Method string
+	Params map[string]interface{}
+}
+
+func readAll(ch <-chan *Class) []*Class {
+	var a []*Class
+	for c := range ch {
+		a = append(a, c)
+	}
+	return a
+}
+
+func main() {
+	var (
+		httpaddr string
+		tlscert  string
+		tlskey   string
+		neoaddr  string
+	)
+
+	flag.StringVar(&httpaddr, "http", "127.0.0.1:8080", "HTTP bind address.")
+	flag.StringVar(&tlscert, "tlscert", "", "Path to TLS certificate.")
+	flag.StringVar(&tlskey, "tlskey", "", "Path to TLS key.")
+	flag.StringVar(&neoaddr, "neo4j", "127.0.0.1:7687", "Neo4j Bolt address.")
+
+	flag.Parse()
+
+	// Initialize the service.
+	svc, err := NewService(neoaddr, -1)
+	if err != nil {
+		log.Fatalf("could not connect to the servce: %s", err)
+	}
+
+	// Top-level context that is cancellable by signals.
+	cxt, cancel := context.WithCancel(context.Background())
+
+	sig := make(chan os.Signal)
+	signal.Notify(sig, os.Interrupt, os.Kill)
+
+	go func() {
+		select {
+		case <-sig:
+			cancel()
+			return
+		}
+	}()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "POST" {
+			w.WriteHeader(405)
+			return
+		}
+
+		var req rpcRequest
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			w.WriteHeader(422)
+			return
+		}
+
+		var (
+			resp interface{}
+			err  error
+		)
+
+		switch req.Method {
+		case "get":
+			vocab := req.Params["vocab"].(string)
+			code := req.Params["code"].(string)
+			resp, err = svc.Get(vocab, code)
+
+		case "validate":
+			vocab := req.Params["vocab"].(string)
+			var codes []string
+			for _, v := range req.Params["codes"].([]interface{}) {
+				codes = append(codes, v.(string))
+			}
+			resp, err = svc.Validate(vocab, codes)
+
+		case "match":
+			vocab := req.Params["vocab"].(string)
+			pattern := req.Params["pattern"].(string)
+			ch := make(chan *Class)
```

Comment:
I can guess why you're using a channel here (to enable cancelling a potentially long running service call?), but it would be nice to get a comment documenting it. Maybe where the context is created on line 52?